fem/transfer.hpp

```diff
@@ -244,6 +244,103 @@ class L2ProjectionGridTransfer : public GridTransfer
+/** The forward, coarse-to-fine, transfer uses L2 projection. The backward,
+    fine-to-coarse, transfer is defined as B = (F^t M_f F)^{-1} F^t M_f, where
+    F is the forward transfer matrix, and M_f is the mass matrix lumped through
+    row-summation. Note that the backward transfer operator, B, is a left
+    inverse of the forward transfer operator, F, i.e. B F = I. Both F and B are
+    defined in physical space.
+
+    This class supports H1 finite element spaces and fine meshes that are a
+    uniform refinement of the coarse mesh.  Generally, the coarse and fine FE
+    spaces can have different orders, however, in order for the backward
+    operator to be well-defined, the number of fine dofs (in a coarse element)
+    should not be smaller than the number of coarse dofs. */
+class L2ProjectionH1GridTransfer : public GridTransfer
+{
+protected:
+   /** Class representing projection operator between a high-order H1 finite
+       element space on a coarse mesh, and a low-order H1 finite element space
+       on a refined mesh (LOR). We assume that the low-order space, fes_lor,
+       lives on a mesh obtained by refining the mesh of the high-order space,
+       fes_ho. */
+   class L2ProjectionH1 : public Operator
+   {
+      const FiniteElementSpace& fes_ho;
+      const FiniteElementSpace& fes_lor;
+
+      SparseMatrix* R;
+      SparseMatrix M_LH;
+      mutable DenseMatrix* P;
+      const double p_rtol;
+      const double p_atol;
+
+      Table ho2lor;
+   public:
+      L2ProjectionH1(const FiniteElementSpace& fes_ho_,
+                     const FiniteElementSpace& fes_lor_);
+      virtual ~L2ProjectionH1();
+      /// Perform the L2 projection onto the LOR space
+      virtual void Mult(const Vector& x, Vector& y) const;
+      /// Perform the transpose of L2 projection onto the LOR space, useful for
+      /// transferring dual fields.
+      virtual void MultTranspose(const Vector& x, Vector& y) const;
+      /// Perform the mass conservative left-inverse prolongation operation.
+      /// This functionality is also provided as an Operator by L2Prolongation.
+      /// Builds the prolongation operator if P is NULL.
+      void Prolongate(const Vector& x, Vector& y) const;
+      /// Perform the transpose of the mass conservative left-inverse
+      /// prolongation operation, useful for transferring dual fields.
+      /// Builds the prolongation operator if P is NULL.
+      void ProlongateTranspose(const Vector& x, Vector& y) const;
+   private:
+      /// Computes sparsity pattern and initializes R matrix.  Based on
+      /// BilinearForm::AllocMat() except maps between HO elements and LOR
+      /// elements.
+      void AllocR();
+      /// Computes prolongation matrix P.
+      void BuildP() const;
+   };
+
+   /** Mass-conservative prolongation operator going in the opposite direction
+       as L2Projection. This operator is a left inverse to the L2Projection. */
```

Comment:
Consider elaborating on "opposite direction", even though I know what you mean.

Reply:
Fixed.